huaweicloud/config/auth.go

```diff
@@ -0,0 +1,374 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// Set UserAgent
+	client.UserAgent.Prepend("terraform-provider-iac")
+
+	config, err := generateTLSConfig(c)
+	if err != nil {
+		return nil, err
+	}
+	transport := &http.Transport{
+		Proxy:           http.ProxyFromEnvironment,
+		TLSClientConfig: config,
+	}
+
+	client.HTTPClient = http.Client{
+		Transport: &LogRoundTripper{
+			Rt:         transport,
+			OsDebug:    logging.IsDebugOrHigher(),
+			MaxRetries: c.MaxRetries,
+		},
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			if client.AKSKAuthOptions.AccessKey != "" {
+				golangsdk.ReSign(req, golangsdk.SignOptions{
+					AccessKey:  client.AKSKAuthOptions.AccessKey,
+					SecretKey:  client.AKSKAuthOptions.SecretKey,
+					RegionName: client.AKSKAuthOptions.Region,
+				})
+			}
+			return nil
+		},
+	}
+
+	if c.MaxRetries > 0 {
+		client.MaxBackoffRetries = uint(c.MaxRetries)
+		client.RetryBackoffFunc = retryBackoffFunc
+	}
+
+	// Validate authentication normally.
+	err = huaweisdk.Authenticate(client, ao)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func genClients(c *Config, pao, dao golangsdk.AuthOptionsProvider) error {
+	client, err := genClient(c, pao)
+	if err != nil {
+		return err
+	}
+	c.HwClient = client
+
+	client, err = genClient(c, dao)
+	if err == nil {
+		c.DomainClient = client
+	}
+	return err
+}
+
+func buildClientByToken(c *Config) error {
+	var pao, dao golangsdk.AuthOptions
+
+	if c.AgencyDomainName != "" && c.AgencyName != "" {
+		pao = golangsdk.AuthOptions{
+			AgencyName:       c.AgencyName,
+			AgencyDomainName: c.AgencyDomainName,
+			DelegatedProject: c.DelegatedProject,
+		}
+
+		dao = golangsdk.AuthOptions{
+			AgencyName:       c.AgencyName,
+			AgencyDomainName: c.AgencyDomainName,
+		}
+	} else {
+		pao = golangsdk.AuthOptions{
+			DomainID:   c.DomainID,
+			DomainName: c.DomainName,
+			TenantID:   c.TenantID,
+			TenantName: c.TenantName,
+		}
+
+		dao = golangsdk.AuthOptions{
+			DomainID:   c.DomainID,
+			DomainName: c.DomainName,
+		}
+	}
+
+	for _, ao := range []*golangsdk.AuthOptions{&pao, &dao} {
+		ao.IdentityEndpoint = c.IdentityEndpoint
+		ao.TokenID = c.Token
+
+	}
+	return genClients(c, pao, dao)
+}
+
+func buildClientByAKSK(c *Config) error {
+	var pao, dao golangsdk.AKSKAuthOptions
+
+	if c.AgencyDomainName != "" && c.AgencyName != "" {
+		pao = golangsdk.AKSKAuthOptions{
+			DomainID:         c.DomainID,
+			Domain:           c.DomainName,
+			AgencyName:       c.AgencyName,
+			AgencyDomainName: c.AgencyDomainName,
+			DelegatedProject: c.DelegatedProject,
+		}
+
+		dao = golangsdk.AKSKAuthOptions{
+			DomainID:         c.DomainID,
+			Domain:           c.DomainName,
+			AgencyName:       c.AgencyName,
+			AgencyDomainName: c.AgencyDomainName,
+		}
+	} else {
+		pao = golangsdk.AKSKAuthOptions{
+			BssDomainID: c.DomainID,
+			BssDomain:   c.DomainName,
+			ProjectName: c.TenantName,
+			ProjectId:   c.TenantID,
```

Comment:
can we change this block to : golangsdk.AuthOptions{
			DomainID:   c.DomainID,
			DomainName: c.DomainName,
			TenantID:   c.TenantID,
			TenantName: c.TenantName,
		}
so this block in buildClientByxxx will be the same